### DISCUSSION TITLE
How to add other boards?

### DISCUSSION THREAD

**Discussion by DuyTrandeLion** [eugene-tarassov/vivado-risc-v#104] [Q&A] [answered]
Hello, I would like to add other boards lile Pynq-Z1 and Ultra96. Do you have any guides for this?

**Comment by mwrnd** [comment, marked as answer]
I am in the process of adding a new board to `vivado-risc-v`. Look at [this comparison](https://github.com/eugene-tarassov/vivado-risc-v/compare/master...mwrnd:vivado-risc-v:innova2) of my branch and the official branch for the changes I have made thus far. I also started on some [notes](https://github.com/mwrnd/innova2_experiments/tree/main/riscv_rocket64b4l2w_xdma#recreating-the-risc-v-design).

First step is to create a Github fork of `vivado-risc-v`. Copy one of the directories under `board` that is closest to your target board and rename it. Edit `board/YOURBOARNAME/Makefile.inc` to the specifics of your board. [XilinxBoardStore](https://github.com/Xilinx/XilinxBoardStore) has support for the [Ultra96](https://github.com/Xilinx/XilinxBoardStore/tree/master/boards/Avnet/ultra96v1/1.2) and some [PYNQ](https://github.com/Xilinx/XilinxBoardStore/tree/master/boards/TUL) boards. If your board does not have an official repo, you can set `BOARD_PART=NONE` which is [now an option](https://github.com/eugene-tarassov/vivado-risc-v/commit/ca7465596ec393024c6531bf2842439a8d283d0d).

Edit `board/YOURBOARNAME/bootrom.dts` to add peripherals to a compiled RISC-V. Ideally you would choose a board to copy that has an identical peripheral set and `bootrom.dts` does not need editing.

You may leave the default frequencies in `board/rocket-freq`.

Edit `ethernet.xdc`, `sdc.xdc`, `top.xdc`, and `uart.xdc` with the pins of your board.

`source` your Vivado's settings in a Terminal and compile a RISC-V project for your customized board. Note that you will need about 20GB of free drive space. `make` will download about 8GB of files on a first compile.
```
source /opt/Xilinx/Vivado/2022.1/settings64.sh
make CONFIG=rocket64b2 BOARD=YOURBOARNAME vivado-tcl
```

If all goes well you will end up with a project under the `workspace` directory that includes a `.tcl` file you can [`source`](https://docs.xilinx.com/r/en-US/ug939-vivado-designing-with-ip-tutorial/Source-the-Tcl-Script?tocId=jFIwPQ0k2AQihKXYcPCJcA) in Vivado. 

Open the Block Design for editing and make any necessary changes. Then [write a new project TCL file](https://docs.xilinx.com/r/2021.1-English/ug895-vivado-system-level-design-entry/Tcl-Command-for-Configuring-Project-Settings) and use the resulting file as `board/YOURBOARNAME/riscv-202?.?.tcl`

It is also useful to compile a project for one of the supported boards to see what happens.

**Reply by DuyTrandeLion** [reply]
Hi @mwrnd , I successfully built for the Arty-A7-100 board which does not have an SD Card slot. May I ask where the code is stored and how should I write a Hello World C program for it? Thanks.

**Comment by mwrnd** [comment]
I do not have an Arty but you should be able to upload and run the `bare-metal` demo using the on-board USB-JTAG. The output of the demo will go to the on-board USB-UART.

Compile the `hello-world` demo and then disassemble it for easier debug.
```
cd vivado-risc-v/bare-metal/hello-world
make
riscv64-unknown-elf-objdump -S -l --inlines -D boot.elf  > boot_elf_dis.txt
```

In one terminal:
```
source /tools/Xilinx/Vivado/[VIVADO_VERSION]/settings64.sh
xsdb

connect
targets
targets 3
stop
dow  vivado-risc-v/bare-metal/hello-world/boot.elf
con
rrd
rr pc
```

![xsdb_bare-metal_hello-world_upload_run](https://user-images.githubusercontent.com/38073881/203917674-1cdf21b6-d730-4c6e-9b70-64e8338e28b0.png)

In a second terminal, run `sudo gtkterm --port /dev/ttyUSB0` to view the output of the `bare-metal` demo. Again, I do not have an Arty so I am guessing that its USB-UART will show up as `/dev/ttyUSB0`. From [this thread](https://forum.digilent.com/topic/3564-problem-with-detecting-arty-on-linux/) it appears the USB-UART may show up as `/dev/ttyUSB1`. Try `ls -l /dev | grep ttyUSB` to see all USB-UART devices.

A picture of a similar task:

![xdma_tty_cuse_Send_RAW_File_test](https://user-images.githubusercontent.com/38073881/203917711-8daa775b-35aa-4735-9881-b198e0ea07e6.png)

**Reply by DuyTrandeLion** [reply]
Hi @mwrnd Thanks for your answer. I have 1 more question before building the hello-world: I am running the rocket32s1 core, so how should I modify the `CFLAGS = -march=rv64imfd -mabi=lp64d -fno-builtin -ffreestanding` to 32-bit?

**Reply by eugene-tarassov** [reply]
```
make CROSS_COMPILE=../../workspace/gcc/riscv/bin/riscv32-unknown-elf- CFLAGS='-march=rv32imac -mabi=ilp32 -fno-builtin -ffreestanding' clean all
```

**Reply by DuyTrandeLion** [reply]
Hi @mwrnd @eugene-tarassov, I used your command and successfully loaded the boot.elf to the board, but there was nothing showing on the uart terminal. Is it because of my RISC-V is misconfigured? I suspect that my DDR3 configuration was not correct.
<img width="1779" alt="image" src="https://user-images.githubusercontent.com/20740534/195756717-416732c6-8cf9-4707-9529-63bccd02adf8.png">

**Reply by mwrnd** [reply]
After compiling the demo for `rocket32s1`, use `riscv32-unknown-elf-objdump` to get a disassembled version to help with debugging:
```
cd vivado-risc-v/bare-metal/hello-world
make CROSS_COMPILE=../../workspace/gcc/riscv/bin/riscv32-unknown-elf- CFLAGS='-march=rv32imac -mabi=ilp32 -fno-builtin -ffreestanding' clean all

../../workspace/gcc/riscv/bin/riscv32-unknown-elf-objdump -S -l --inlines -D boot.elf   >  dis.txt
```

`rr` is *Read Register [ALL]* and `rr pc` is *Read Register Program Counter*. It is at `0x8000033e` in your image which means your `rocket32s1` is running inside the delay loop. It seems to be working.
![hello-world_boot elf_disassembled](https://user-images.githubusercontent.com/38073881/195897750-daf86e5e-27db-496e-b0df-e15ea1046338.png)

If you are running Fedora, it has [`gtkterm`](https://packages.fedoraproject.org/pkgs/gtkterm/gtkterm/). Please install and try it on all the USB-UART ports. I know its defaults work for me. One less variable in the mix.
```
sudo gtkterm --speed 115200 --port /dev/ttyUSB0
sudo gtkterm --speed 115200 --port /dev/ttyUSB1
```

Also, give `hello-world` more time. It it supposed to sleep for [1 second](https://github.com/eugene-tarassov/vivado-risc-v/blob/36b5433909945d1a34a0e367f55164ae1335f6f9/bare-metal/hello-world/main.c#L21) in between its UART writes but if there is some misconfiguration it may take longer.

Next thing to try would be reading memory using `xsdb` to establish whether the RISC-V core is communicating with the UART. The Nexys-A7 has its [UART at `0x60010000`](https://github.com/eugene-tarassov/vivado-risc-v/blob/36b5433909945d1a34a0e367f55164ae1335f6f9/board/nexys-a7-100t/bootrom.dts#L17). The [Status Register is at +`0x08`](https://github.com/eugene-tarassov/vivado-risc-v/blob/36b5433909945d1a34a0e367f55164ae1335f6f9/uart/uart.v#L283). Try `mrd 0x60010008` in the `xsdb` console.
```
connect
targets
target 3
stop
dow  vivado-risc-v/bare-metal/hello-world/boot.elf
con
after 1000
mrd -size b 0x60010008 1
```

The `Status` register bits are defined as `4=CTS, 3=TX_Full, 2=TX_Empty, 1=RX_Full, 0=RX_Has_Data`. For example, if you get `0x14` you can run `echo -n -e '\x14' | xxd -b` in a terminal which will display `00010100`which means `Clear-To-Send` and `TX_Empty` are true. Run `mrd 0x60010008` several times to see if you can catch any bytes being transferred and whether the TX FIFO is ever full. If `TX_Full` is true and stays true then nothing is reading the data that the demo is generating. If CTS is false then the UART cannot send data. The [Arty-A7](https://digilent.com/reference/_media/arty:arty_sch.pdf) uses only RX and TX pins while the [Nexys-A7](https://digilent.com/reference/_media/reference/programmable-logic/nexys-4-ddr/nexys_a7_sch-public.pdf) includes CTS and RTS connections. Try connecting the `CTSn` pin of the UART in your Vivado Block Design to Constant with Value 0.

If the above does not work, please post a picture of your Block Design from Vivado. Let us look over your design before you [add an AXI ILA](https://docs.xilinx.com/r/en-US/ug994-vivado-ip-subsystems/Using-the-ILA-IP-to-Debug-a-Block-Design). You can use [`write_bd_layout -format pdf -orientation landscape MYBLOCKDESIGN.pdf`](https://docs.xilinx.com/r/en-US/ug835-vivado-tcl-commands/write_bd_layout) in the Vivado Tcl Console to generate a PDF. [GhostScript](https://packages.fedoraproject.org/pkgs/ghostscript/ghostscript/) can then [be used](https://stackoverflow.com/questions/653380/converting-a-pdf-to-png) in a terminal to convert the PDF to a PNG image: `gs -sDevice=png16m -sOutputFile="MYBLOCKDESIGN.png" MYBLOCKDESIGN.pdf`. Please use [ImageMagick](https://packages.fedoraproject.org/pkgs/ImageMagick/ImageMagick/) to resize the PNG to something that is readable but [as small an image as possible](https://askubuntu.com/questions/871218/resize-images-keep-ratio-with-size-constraints). Try: `mogrify -resize '1000x1000^' MYBLOCKDESIGN.png`.

**Reply by DuyTrandeLion** [reply]
After getting the `CTSn` pin to constant 0, I got the example working. I think the same method should be applied for `RSTn`, is it right?
Last question: Can I contribute the Arty-A7 project to this repo? If yes, do you need the include the SD Card to the design and require an [external pmod to micro sdcard](https://digilent.com/shop/pmod-microsd-microsd-card-slot/) module to run Linux?

**Reply by mwrnd** [reply]
`RTSn` is an output from the UART and can be left floating. The [Arty-A7](https://digilent.com/reference/_media/arty:arty_sch.pdf) only uses the RX and TX pins.

As for contributing, wait for @eugene-tarassov to respond but please [fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/about-forks) the repo either way in case other Arty-A7 users want to give it a go. Please mention the Arty-A7 in the *About* description so it is easier to search for on GitHub.

Note that @eugene-tarassov [mentioned](https://github.com/eugene-tarassov/vivado-risc-v/issues/97#issuecomment-1205602934) in a previous post about having a "fully automated build and test system running continuously with multiple target boards". Adding another board may be too much of a support hassle.

**Reply by eugene-tarassov** [reply]
> Can I contribute the Arty-A7 project to this repo?

Yes. It would be nice. The recommended way is to fork the repo and create pull request.

> do you need the include the SD Card to the design

I would prefer the SD card controller included, and Linux supported with SD card pmod, but it is not required.

**Reply by eugene-tarassov** [reply]
I have not heard from you for a while, so I have added support for Arty A7 100T

**Reply by DuyTrandeLion** [reply]
Hi Eugene, I was stuck with the MII interface. It looks like you have added it... Sorry I did not make it in time.
